src/monitor/group_state_machine.c

```diff
@@ -570,6 +885,591 @@ ProceedGroupStateForPrimaryNode(AutoFailoverNode *primaryNode)
+			activeNode->nodeId,
+			activeNode->nodeHost,
+			activeNode->nodePort,
+			candidateNode->nodeId,
+			candidateNode->nodeHost,
+			candidateNode->nodePort);
+
+		AssignGoalState(activeNode, REPLICATION_STATE_JOIN_SECONDARY, message);
+
+		return true;
+	}
+
+	/* when we have a candidate, we don't go through finding a candidate */
+	return false;
+}
+
+
+/*
+ * SelectFailoverCandidateNode returns the candidate to failover to when we
+ * have one already.
+ *
+ * The selection is based on candidatePriority. If the candidate with the
+ * higher priority doesn't have the most recent LSN, we have it fetch the
+ * missing WAL bits from one of the standby which did receive them.
+ *
+ * Before we enter the selection process, we must have collected the last
+ * received LSN from ALL the standby nodes that are considered as a candidate
+ * (thanks to the FSM transition secondary -> report_lsn), and now we need to
+ * select one of the failover candidates.
+ *
+ * As input we get the candidateNodesGroupList, a filtered list of standby that
+ * are known to be a failover candidate from an earlier filtering process. We
+ * also get the mostAdvancedNode and the primaryNode so that we can decide on
+ * the next step (cascade WALs or promote directly).
+ */
+static AutoFailoverNode *
+SelectFailoverCandidateNode(CandidateList *candidateList,
+							AutoFailoverNode *primaryNode)
+{
+	/* build the list of failover candidate nodes, ordered by priority */
+	List *sortedCandidateNodesGroupList =
+		GroupListCandidates(candidateList->candidateNodesGroupList);
+
+	/* it's only one of the most advanced nodes, a reference to compare LSN */
+	AutoFailoverNode *mostAdvancedNode =
+		(AutoFailoverNode *) linitial(candidateList->mostAdvancedNodesGroupList);
+
+	/* the goal in this function is to find this one */
+	AutoFailoverNode *selectedNode = NULL;
+
+	ListCell *nodeCell = NULL;
+
+	/*
+	 * We refuse to orchestrate a failover that would have us lose more data
+	 * than is configured on the monitor. Both when using sync and async
+	 * replication we have the same situation that could happen, where the most
+	 * advanced standby node in the system is lagging behind the primary and
+	 * promoting it would incur data loss.
+	 *
+	 * In sync replication, that happens when the primary has been waiting for
+	 * a large chunk of WAL bytes to be reported. In async, the only difference
+	 * is that the primary did not wait.
+	 *
+	 * In terms of client-side guarantees, it's a big difference. In term of
+	 * data durability, it's the same thing.
+	 *
+	 * For this situation to change, users will have to either re-live the
+	 * unhealthy primary or change the
+	 * pgautofailover.enable_sync_wal_log_threshold GUC to a larger value and
+	 * thus explicitely accept data loss.
+	 */
+	if (primaryNode &&
+		!WalDifferenceWithin(mostAdvancedNode, primaryNode, PromoteXlogThreshold))
+	{
+		char message[BUFSIZE] = { 0 };
+
+		LogAndNotifyMessage(
```

Comment:
Minor/optional: It might be nice to hint users that they can increase `pgautofailover.enable_sync_wal_log_threshold` to allow failover. It might not be obvious for users I guess

Reply:
Well the idea to spell the GUC in its full name is exactly that, hinting the user that they can change it. And then they can even read the docs about it ;-) I think we're good that way, even though I see your point. But there's only so much we can do in an error message...